src/dc_tools.rs

```diff
@@ -1375,6 +1395,16 @@ pub fn listflags_has(listflags: u32, bitindex: usize) -> bool {
     (listflags & bitindex) == bitindex
 }
 
+/// Macro to convert strings  to `const char *`. It can't be implemented
+/// as plain function due the way of how Rust scoping rules work.
+#[macro_export]
+macro_rules! cptr {
+    ($e:expr) => {{
+        use crate::dc_tools::CStringExt;
+        std::ffi::CString::yolo($e).as_ptr() as *const libc::c_char
```

Comment:
Is this safe?  IIRC at some point we refactored all calls where something like `foo(CString::yolo("hey").as_ptr())` was used because this is not safe as the string **may** already have been deallocated by the time the pointer is used and gave us some random crashes.  I'm not entirely sure on the scopes involved with macros, but I'm far from sure the pointer returned here would stay alive long enough.

Reply:
cc @dignifiedquire

Reply:
I did some experimenting, that was meant to emulate situation before writing this macro:

``
struct Bar {
    pub x: u32
}

impl Bar {
    pub fn new(x: u32) -> Self {
        println!("New Bar made!");
        Bar { x }
    }
    pub fn extract(&self) -> u32{
        self.x
    }
}

impl Drop for Bar {
    fn drop(&mut self) {
        println!("Bar dropped!");
    }
}

pub fn foo(arg: u32) {
    println!("Function called with arg = {}", arg);
}

fn main () {
    foo(Bar::new(12).extract());
}
``

Here, `Bar` is dropped after function returns. I will try to find something more solid.

Reply:
According to [this](https://doc.rust-lang.org/reference/expressions.html#temporary-lifetimes), lifetime of temporary object is end of innermost **statement**, which is exactly what we need.

Or I read the reference wrong?

Reply:
Here is [SO thread](https://stackoverflow.com/questions/41815619/are-raw-pointers-to-temporaries-ok-in-rust) that claims that temporary objects are dropped after function call.

Reply:
> Is this safe? IIRC at some point we refactored all calls where something like `foo()` was used because this is not safe as the string **may** already have been deallocated by the time the pointer is used and gave us some random crashes. I'm not entirely sure on the scopes involved with macros, but I'm far from sure the pointer returned here would stay alive long enough.

I think the problem was with code similar to this:
```
let tmp = CString::yolo("hey").as_ptr();
foo(tmp);
```
and the safe alternative is
```
let tmp = CString::yolo("hey");
foo(tmp.as_ptr());
```
so CString holds the memory pointed to by `as_ptr` until `foo` returns.

Reply:
Macros don't affect lifetimes. First, macros are expanded, then the code compiles as usual.

Looks like current usage of this macro is safe, but it is very easy to misuse it, like this:
```
let tmp = cptr!("hi");
foo(tmp);
```

Reply:
We do it like this:

```
fn foo(x: *const libc::c_char);
let arg = "bar";
with_cptr(arg, |arg| foo(arg))
```

This is how Haskell does it (so-called bracket pattern). But it is quite verbose.

Reply:
Yes, that's probably too verbose for transition macro that will be removed eventually (after mmime rewrite) anyway.

To answer @flub question, it's safe as long as used properly.

Reply:
The macro as it is written is not safe, as you introduce a new scope using the double curly brace `{{ }}`. 

I think we should go for a with_ pattern here, this is too easy to get wrong, even if it is temporary. 

The macro expands to 

```rust
{
    use crate::dc_tools::CStringExt;
    std::ffi::CString::yolo($e).as_ptr() as *const libc::c_char
}
```
Which means the `CString` is dropped before it can be used by whatever call site is using the macro, e.g.

```rust
call_fn(cptr!("hello"));

// becomes

call_fn({
  use crate::dc_tools::CStringExt;
  std::ffi::CString::yolo("hello").as_ptr() as *const libc::c_char
});
```

Reply:
@dignifiedquire Seems like you are right, the innermost expression is a [block expression](https://doc.rust-lang.org/reference/expressions/block-expr.html), and the variable lifetime is limited by it.

@KAction Maybe just keep explicit `CString::yolo` calls where `with_` is too verbose.

Reply:
Okay, this macro was bad idea. Will drop and bring `_c` convention back.